netket/graph/_lattice_edge_logic.py

```diff
@@ -0,0 +1,153 @@
+    row, col, dst = find(triu(dist_matrix))
+    dst = comparable(dst)
+    _, ii = np.unique(dst, return_inverse=True)
+
+    return [sorted(list(zip(row[ii == k], col[ii == k]))) for k in range(order)]
+
+
+def get_nn_edges(
+    basis_vectors,
+    extent,
+    site_offsets,
+    pbc,
+    distance_atol,
+    order,
+):
+    positions, ids = create_padded_sites(
+        basis_vectors, extent, site_offsets, pbc, order
+    )
+    naive_edges_by_order = get_naive_edges(
+        positions,
+        order * np.linalg.norm(basis_vectors, axis=1).max() + distance_atol,
+        order,
+    )
+    colored_edges = []
+    for k, naive_edges in enumerate(naive_edges_by_order):
+        true_edges = set()
+        for node1, node2 in naive_edges:
+            # switch to real node indices
+            node1 = ids[node1]
+            node2 = ids[node2]
+            if node1 == node2:
+                raise RuntimeError(
+                    f"Lattice contains self-referential edge {(node1, node2)} of order {k}"
+                )
+            elif node1 > node2:
+                node1, node2 = node2, node1
+            true_edges.add((node1, node2))
+        for edge in true_edges:
+            colored_edges.append((*edge, k))
+    return colored_edges
+
+
+# Unit cell distribution logic
+
+
+def get_custom_edges(basis_vectors, extent, site_offsets, pbc, atol, descriptor):
+    def translated_edges(sl1, sl2, distance, color):
+        # get distance in terms of unit cells
+        d_cell = (distance + site_offsets[sl1] - site_offsets[sl2]) @ np.linalg.inv(
+            basis_vectors
+        )
+
+        if not np.all(is_approx_int(d_cell, atol=atol)):
+            # error out
+            msg = f"{distance} is invalid distance vector between sublattices {sl1}->{sl2}"
+            # see if the user flipped the vector accidentally
+            d_cell = (distance + site_offsets[sl2] - site_offsets[sl1]) @ np.linalg.inv(
+                basis_vectors
+            )
+            if np.all(is_approx_int(d_cell, atol=atol)):
+                msg += f" (but valid {sl2}->{sl1})"
+            raise ValueError(msg)
+
+        d_cell = np.asarray(np.rint(d_cell), dtype=int)
+
+        # catches self-referential and other unrealisably long edges
+        if not np.all(d_cell < extent):
+            raise ValueError(
+                f"Distance vector {distance} does not fit into the lattice"
+            )
+
+        # Unit cells of starting points
+        start_min = np.where(pbc, 0, np.maximum(0, -d_cell))
+        start_max = np.where(pbc, extent, extent - np.maximum(0, d_cell))
+        start_ranges = [slice(lo, hi) for lo, hi in zip(start_min, start_max)]
+        start = np.mgrid[start_ranges].reshape(len(extent), -1).T
+        end = (start + d_cell) % extent
+
+        # Convert to site indices
+        radix = np.cumprod([len(site_offsets), *extent[:0:-1]])[::-1]
+        start = start @ radix + sl1
+        end = end @ radix + sl2
```

Comment:
Maybe this can also be extracted into a helper function, given that it is very similar to the id generating code in `create_padded_sites`.